pkg/common/cns-lib/volume/listview.go

```diff
@@ -0,0 +1,328 @@
+	return nil
+}
+
+// listenToTaskUpdates is a long-running goroutine
+// that uses a property collector to listen for task updates
+// CSI ops add CNS tasks to listview and wait for a response from CNS
+// when update(s) are received by the property collector,
+// it spawns a new goroutine to process each task update and return the result to the caller
+func (l *ListViewImpl) listenToTaskUpdates() {
+	log := logger.GetLogger(l.ctx)
+	filter := getListViewWaitFilter(l.listView)
+	// we need to recreate the listView and the wait filter after any error from vc
+	// for the first iteration we already have the listView and filter initialized
+	recreateView := false
+	for {
+		// calling Connect at the beginning to ensure the current session is neither nil nor NotAuthenticated
+		if err := l.isClientValid(); err != nil {
+			log.Errorf("failed to connect to vCenter. err: %v", err)
+			time.Sleep(waitForUpdatesRetry)
+			continue
+		} else {
+			log.Infof("connection to vc successful")
+		}
+
+		if recreateView {
+			log.Info("re-creating the listView object")
+			err := l.createListView(l.ctx, nil)
+			if err != nil {
+				log.Errorf("failed to create a ListView. error: %+v", err)
+				continue
+			}
+
+			filter = getListViewWaitFilter(l.listView)
+			recreateView = false
+		}
+
+		log.Info("Starting listening for task updates...")
+		pc := property.DefaultCollector(l.govmomiClient.Client)
+		err := property.WaitForUpdates(l.ctx, pc, filter, func(updates []types.ObjectUpdate) bool {
+			log.Debugf("Got %d property collector update(s)", len(updates))
+			for _, update := range updates {
+				for _, prop := range update.ChangeSet {
+					log.Debugf("Got update for object %v properties %v", update.Obj, prop)
+					// we don't need a lock at this line as we aren't accessing any map item
+					go l.processTaskUpdate(prop)
+				}
+			}
+
+			// this return value is used by the WaitForUpdates method.
+			// we only want this true while running the unit tests so the test can finish
+			return l.shouldStopListening
+		})
+		// if property collector returns any errors,
+		// we want to immediately return a fault for all the pending tasks in the map
+		// note: this is not a task error but an error from the vc
+		if err != nil {
+			log.Errorf("WaitForUpdates returned err: %v for vc: %+v", err, l.virtualCenter)
+			recreateView = true
+
+			// in case of failure, we return failure for the tasks present in the map
+			// reading from a map requires acquiring a reader lock
+			// if a writer lock is called first, reader lock will block until the writer lock is released & vice-versa
+
+			l.taskMap.GetAll()
+			for _, taskDetails := range l.taskMap.GetAll() {
+				prop := types.PropertyChange{
+					Val: types.TaskInfo{
+						// Task:  taskMoRef,
+						Task:  taskDetails.Reference,
+						State: types.TaskInfoStateError,
+						Error: &types.LocalizedMethodFault{
+							LocalizedMessage: err.Error(),
+						},
+					},
+				}
+				go l.processTaskUpdate(prop)
```

Comment:
Isn't this repetitive? You pick a lock and get all the tasks from taskMap, followed by creating a goroutine, which again picks a lock and get the task details from taskMap. Do we need multiple goroutines doing this? is there a possibility of batching the result for mark-all-fail scenarios? Something similar to `RemoveTasksMarkedForDeletion` , you can have `ReportErrorOnAllTasks`

Reply:
Fixed.